### DISCUSSION TITLE
`Unit` is not inferred as `INothing`

### DISCUSSION THREAD

**Discussion by Fristi** [typelevel/fs2#2721] [Q&A] [answered]
I have a `Pipe[F, Byte, Unit]` but this is not inferred as `Pipe[F, Byte, INothing]` ? I assume that's the case, is there a combinator for this cast?

**Comment by mpilquist** [comment, marked as answer]
Use the `.drain` method on `Stream` to discard those units. For example:
```scala
val p: Pipe[F, Byte, Unit] = ???
val q: Pipe[F, Byte, Nothing] = p.andThen(_.drain)
```

That discarding has to be manual. To see why, consider this program:

```scala
val s: Stream[Pure, Unit] = Stream(1, 2, 3).as(())
val s2: Stream[Pure, Int] = Stream(4, 5, 6)
val s3 = s ++ s2
```
What's the type of `s3` here? It's a `Stream[Pure, AnyVal]` because `AnyVal` is the least upper bound of `Unit` and `Int`.

**Reply by Fristi** [reply]
That works! Thanks!
